sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT);
+      reg = (reg & ~0x03ff) | var;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg);
+      return 0;
+    case EMMC_PROP_RW_RESP_LOW:
+      b_out(emmc, BCMEMMC_RESP0, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP1, (uint32_t)(var >> 32));
+      return 0;
+    case EMMC_PROP_RW_RESP_HI:
+      b_out(emmc, BCMEMMC_RESP2, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP3, (uint32_t)(var >> 32));
+      return 0;
+    case EMMC_PROP_RW_CLOCK_FREQ:
+      return bcmemmc_clk(cdev->parent, var);
+    case EMMC_PROP_RW_BUSWIDTH:
+      return bcmemmc_set_bus_width(state, var);
+    case EMMC_PROP_RW_RCA:
+      state->rca = var;
+      return 0;
+    default:
+      return ENODEV;
+  }
+}
+
+static int bcmemmc_cmd_code(device_t *dev, uint32_t code, uint32_t arg,
+                            emmc_resp_t *resp) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+
+  uint32_t r = 0;
+
+  b_out(emmc, BCMEMMC_ARG1, arg);
+  b_out(emmc, BCMEMMC_CMDTM, code);
+  if ((r = bcmemmc_intr_wait(dev, INT_CMD_DONE))) {
+    klog("ERROR: failed to send EMMC command %p", code);
+    return r;
+  }
+
+  if (resp) {
+    resp->r[0] = b_in(emmc, BCMEMMC_RESP0);
+    resp->r[1] = b_in(emmc, BCMEMMC_RESP1);
+    resp->r[2] = b_in(emmc, BCMEMMC_RESP2);
+    resp->r[3] = b_in(emmc, BCMEMMC_RESP3);
+  }
+
+  return 0;
+}
+
+static int bcmemmc_cmd(device_t *cdev, emmc_cmd_t cmd, uint32_t arg,
+                       emmc_resp_t *resp) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+
+  if (cmd.flags & EMMC_F_APP)
+    bcmemmc_cmd(cdev, EMMC_CMD(APP_CMD), state->rca << 16, NULL);
+
+  uint32_t code = encode_cmd(cmd);
+  return bcmemmc_cmd_code(cdev->parent, code, arg, resp);
+}
+
+static int bcmemmc_read(device_t *cdev, void *buf, size_t len, size_t *read) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  device_t *emmcdev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)emmcdev->state;
+  resource_t *emmc = state->emmc;
+
+  assert(is_aligned(len, 4)); /* Assert multiple of 32 bits */
+
+  /* A very simple transfer */
+  for (size_t i = 0; i < len / sizeof(uint32_t); i++)
+    ((uint32_t *)buf)[i] = b_in(emmc, BCMEMMC_DATA);
+
+  /* TODO (mohr): check wether the transfer fully succeeded! */
+  if (read)
+    *read = len;
+  return 0;
+}
+
+static int bcmemmc_write(device_t *cdev, const void *buf, size_t len,
+                         size_t *wrote) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  device_t *emmcdev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)emmcdev->state;
+  resource_t *emmc = state->emmc;
+
+  assert(is_aligned(len, 4)); /* Assert multiple of 32 bits */
+
+  /* A very simple transfer */
+  for (size_t i = 0; i < len / sizeof(uint32_t); i++)
+    b_out(emmc, BCMEMMC_DATA, ((uint32_t *)buf)[i]);
```

Comment:
This looks kinda inconsistent since you lose `const`.

Reply:
Fixed.